### PR TITLE
[RHELC-1070] Duplicate package check for preconversion analysis

### PR DESCRIPTION
**Pull request by pr-watson**
This PR adds a check to preconversion analysis to run "package-cleanup --dupes" to see if the users system contains multiple versions of the same package. If multiple versions of the same package are detected then an error is raised and the user is instructed to remove the duplicate packages with the help of the package-cleanup command. 

Jira Issues: [RHELC-1070](https://issues.redhat.com/browse/RHELC-1070)

Checklist

- [x] PR has been tested manually in a VM (either author or reviewer)
- [x] Jira issue has been made public if possible
- [x] `[RHELC-]` is part of the PR title 
- [x] GitHub label has been added to help with Release notes 
- [x] PR title explains the change from the user's point of view
- [x] Code and tests are documented properly
- [x] The commits are squashed to as few commits as possible (without losing data) 
- [ ] When merged: Jira issue has been updated to `Release Pending` if relevant

### CODE REVIEW

File: convert2rhel/actions/system_checks/duplicate_packages.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+import logging
+
+from convert2rhel import actions, utils
+
+
+logger = logging.getLogger(__name__)
+
+
+class Duplicate_Packages(actions.Action):
```
Comment (r0x0d) [inline convert2rhel/actions/system_checks/duplicate_packages.py:26]:
```suggestion
class DuplicatePackages(actions.Action):
```

File: convert2rhel/actions/system_checks/duplicate_packages.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    id = "DUPLICATE_PACKAGES"
+
+    def run(self):
+        """Ensure that there are no duplicate system packages installed."""
+        super(Duplicate_Packages, self).run()
+
+        logger.task("Prepare: Check if there are any duplicate installed packages on the system")
+
+        output, _ = utils.run_subprocess(["package-cleanup", "--dupes", "--quiet"], print_output=False)
```
Comment (r0x0d) [inline convert2rhel/actions/system_checks/duplicate_packages.py:35]: Would be best to use the full path to the binary

File: convert2rhel/actions/system_checks/duplicate_packages.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+class DuplicatePackages(actions.Action):
+    id = "DUPLICATE_PACKAGES"
+
+    def run(self):
+        """Ensure that there are no duplicate system packages installed."""
+        super(DuplicatePackages, self).run()
+
+        logger.task("Prepare: Check if there are any duplicate installed packages on the system")
+
+        output, _ = utils.run_subprocess(["/usr/bin/    package-cleanup", "--dupes", "--quiet"], print_output=False)
```
Comment (r0x0d) [inline convert2rhel/actions/system_checks/duplicate_packages.py:35]: Is the space here intentional?

Comment (pr-watson) [reply convert2rhel/actions/system_checks/duplicate_packages.py:35]: Would look better without it, I'll remove

Comment (SpyTec) [reply convert2rhel/actions/system_checks/duplicate_packages.py:34]:
```suggestion
        output, _ = utils.run_subprocess(["/usr/bin/package-cleanup", "--dupes", "--quiet"], print_output=False)
```

**Review by SpyTec — CHANGES_REQUESTED** [review]

File: convert2rhel/unit_tests/actions/system_checks/duplicate_packages_test.py
Comment (SpyTec) [inline convert2rhel/unit_tests/actions/system_checks/duplicate_packages_test.py:1]: Please also expand unit test to when there are no duplicate packages

Comment (pr-watson) [reply convert2rhel/unit_tests/actions/system_checks/duplicate_packages_test.py:1]: Will do thanks

File: convert2rhel/actions/system_checks/duplicate_packages.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        """Ensure that there are no duplicate system packages installed."""
+        super(DuplicatePackages, self).run()
+
+        logger.task("Prepare: Check if there are any duplicate installed packages on the system")
+        output, _ = utils.run_subprocess(["/usr/bin/package-cleanup", "--dupes", "--quiet"], print_output=False)
+        if not output:
+            return
+
+        duplicate_packages = output.split("\n")
+        duplicate_packages = filter(None, duplicate_packages)
```
Comment (r0x0d) [inline convert2rhel/actions/system_checks/duplicate_packages.py:39]: We can simplify the statement by passing the output to the filter directly.

```suggestion
        duplicate_packages = filter(None, output.split("\n"))
```

File: convert2rhel/actions/system_checks/duplicate_packages.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+class DuplicatePackages(actions.Action):
+    id = "DUPLICATE_PACKAGES"
+
+    def run(self):
+        """Ensure that there are no duplicate system packages installed."""
+        super(DuplicatePackages, self).run()
+
+        logger.task("Prepare: Check if there are any duplicate installed packages on the system")
+        output, _ = utils.run_subprocess(["/usr/bin/package-cleanup", "--dupes", "--quiet"], print_output=False)
```
Comment (r0x0d) [inline convert2rhel/actions/system_checks/duplicate_packages.py:34]: Also, the package `package-cleanup` is part of the `dnf-utils` package. I just saw that in our specfile we have a comment saying we require dnf-utils because of yumdownloader. I think it would be nice if updated the comment there and specify this package name as well.

File: convert2rhel/actions/system_checks/duplicate_packages.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        if not output:
+            return
```
[the remark this answers is not in the document]
Comment (r0x0d) [reply convert2rhel/actions/system_checks/duplicate_packages.py:36]: Ah, I see... It returns 0 even if there is no duplicate packages or if it errors out.

[the remark this answers is not in the document]
Comment (pr-watson) [reply convert2rhel/actions/system_checks/duplicate_packages.py:36]: Yes, by checking if there is output we can determine if there are duplicates are not. With the --quiet option if there are no duplicates there will be no output

**Review by r0x0d — COMMENTED** [review]
Few comments, but otherwise, looks good!

File: convert2rhel/actions/system_checks/duplicate_packages.py (hunk 5 above)
Comment (pr-watson) [reply convert2rhel/actions/system_checks/duplicate_packages.py:34]: I'll add that!

Comment (SpyTec) [reply convert2rhel/actions/system_checks/duplicate_packages.py:34]: Did we add this? Can be done in a different PR

File: convert2rhel/actions/system_checks/duplicate_packages.py (hunk 4 above)
Comment (SpyTec) [reply convert2rhel/actions/system_checks/duplicate_packages.py:39]: I'd use the simplification here as well @pr-watson

**Review by SpyTec — APPROVED** [review]
LGTM, just minor things. Can be corrected after test runs are done

File: convert2rhel/actions/system_checks/duplicate_packages.py (hunk 5 above)
Comment (pr-watson) [reply convert2rhel/actions/system_checks/duplicate_packages.py:34]: Yes I made the addition already @SpyTec

**Comment by pr-watson** [comment]
With the Ethel Stage refresh going on today all tests here will be failing

**Comment by pr-watson** [comment]
Once https://github.com/oamg/convert2rhel/pull/1095 is merged the integration test failures related to the redhatrelease file backup should be fixed
